### PR TITLE
[SRC] one controller only

### PR DESCRIPTION
**Pull request by Lironrad**

### CODE REVIEW

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 1]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+    //                }
+    //            }
+    //            //there is no controller for the given group in redis, should take ownership
+    //            else {
+    //                handleGroupManagedByThisController(cloud, controllerIdentifier);
+    //            }
+    //        }
+    //    }
+
+    public void syncGroupsOwner(BaseSpotinstCloud cloud) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: dont need to send cloud

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 2]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+        String groupId   = cloud.getGroupId();
+        String accountId = cloud.getAccountId();
+        LOGGER.info(String.format("try fetching controller identifier for group %s from redis", groupId));
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: Should the key be of the controller or the cloud, when this ctor been called? only once or after each update, does it prevent case that the group define doublicated in the same controller

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 3]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+
+        IRedisRepo          redisRepo             = RepoManager.getInstance().getRedisRepo();
+        ApiResponse<Object> redisGetValueResponse = redisRepo.getValue(groupId, accountId);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: rename LockRepo, the value should be for string only

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 4]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+                        }
+                    }
+                }
+            }
+            else {
+                LOGGER.warn("redis response value return null");
+            }
+        }
+        //there is no controller for the given group in redis, should take ownership
+        else {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:843]: This should be for failure during reading from redis only, just write the error and the current status to log and stay in the current status

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java [hunk 5]
```diff
@@ -313,6 +317,75 @@ public static Boolean azureVmDetach(String groupId, String vmId, String accountI
...
+    //TODO Liron - change name to getLock
+    public static <T> T getRedisValue(String groupId, String accountId) throws ApiException {
+        T retVal      = null;
+
+        Map<String, String>    headers     = buildHeaders();
+
+        Map<String, String>    queryParams = buildQueryParams(accountId);
+
+        RestResponse response =
+                RestClient.sendGet(SPOTINST_API_HOST + "/aws/ec2/group/" + groupId + "/jenkinsPlugin", headers, queryParams);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: Add the suffix "lock" to the url (and change the optimizer)

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java [hunk 6]
```diff
@@ -313,6 +317,75 @@ public static Boolean azureVmDetach(String groupId, String vmId, String accountI
...
+
+        RedisSetKeyRequest request = new RedisSetKeyRequest();
+        request.setGroupId(lockKey);
+        request.setControllerIdentifier(lockValue);
+        request.setTtl(ttl);
+
+        String body = JsonMapper.toJson(request);
+
+        RestResponse response =
+                RestClient.sendPost(SPOTINST_API_HOST + "/aws/ec2/group/jenkinsPlugin", body, headers, queryParams);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: Add the suffix "lock" to the url (and change the optimizer)

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java [hunk 7]
```diff
@@ -313,6 +317,75 @@ public static Boolean azureVmDetach(String groupId, String vmId, String accountI
...
+    //TODO Liron - change name to unlock
+    public static Integer deleteRedisKey(String groupId, String accountId) throws ApiException {
+        Integer retVal      = null;
+
+        Map<String, String>    headers     = buildHeaders();
+
+        Map<String, String>    queryParams = buildQueryParams(accountId);
+
+        RestResponse response =
+                RestClient.sendDelete(SPOTINST_API_HOST + "/aws/ec2/group/" + groupId + "/jenkinsPlugin", headers, queryParams);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: Add the suffix "lock" to the url (and change the optimizer)

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 8]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+        String accountId = cloud.getAccountId();
+        LOGGER.info(String.format("try fetching controller identifier for group %s from redis", groupId));
+
+        IRedisRepo          redisRepo             = RepoManager.getInstance().getRedisRepo();
+        ApiResponse<Object> redisGetValueResponse = redisRepo.getValue(groupId, accountId);
+        String              controllerIdentifier  = SpotinstContext.getInstance().getControllerIdentifier();
+
+        if (redisGetValueResponse.isRequestSucceed()) {
+            //redis response might return in different types
+            if (redisGetValueResponse.getValue() instanceof String) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: need to change null, isGroupLocked = redisGetValueResponse.getValue() == null

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 9]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+                        handleGroupManagedByThisController(cloud, controllerIdentifier);
+                    }
+                    else {
+                        LOGGER.info(String.format("group %s does not belong to controller with identifier %s", groupId,
+                                                  controllerIdentifier));
+                        GroupStateTracker groupStateDetails =
+                                SpotinstContext.getInstance().getConnectionStateByGroupId().get(groupId);
+
+                        if (groupStateDetails == null) {
+                            groupStateDetails = new GroupStateTracker();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: Do it in the ctor

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 10]
```diff
@@ -643,6 +732,179 @@ protected Integer getPendingThreshold() {
...
+                            groupStateDetails.setGroupId(groupId);
+                            groupStateDetails.setAccountId(accountId);
+                            groupStateDetails.setState(SPOTINST_CLOUD_COMMUNICATION_INITIALIZING);
+                            Date now = new Date();
+                            groupStateDetails.setTimeStamp(now);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: Do it internally

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 1 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 4 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:843]: done

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java (hunk 5 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: finished in plugin, optimizer and core service remains

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java (hunk 6 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: done

File: src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java (hunk 7 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/api/SpotinstApi.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 8 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 9 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 10 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 11]
```diff
@@ -110,6 +116,10 @@ public BaseSpotinstCloud(String groupId, String labelString, String idleTerminat
         else {
             this.globalExecutorOverride = new SpotGlobalExecutorOverride(false, 1);
         }
+
+        if (StringUtils.isNotEmpty(groupId) && StringUtils.isNotEmpty(accountId)) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:120]: Consider to move from ctor

Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:120]: boolean isActiveCloud = ...

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 12]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+    //            //there is no controller for the given group in redis, should take ownership
+    //            else {
+    //                handleGroupManagedByThisController(cloud, controllerIdentifier);
+    //            }
+    //        }
+    //    }
+
+    public void syncGroupsOwner() {
+        ILockRepo           lockRepo                    = RepoManager.getInstance().getLockRepo();
+        ApiResponse<String> lockGroupControllerResponse = lockRepo.getLockValueById(groupId, accountId);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:820]: rename to getGroupControllerLock, lockGroupController unlockGroupController

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 13]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+        String              controllerIdentifier        = SpotinstContext.getInstance().getControllerIdentifier();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:821]: move to the if block

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 14]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+                }
+                else {
+                    LOGGER.info(String.format("group %s does not belong to controller with identifier %s", groupId,
+                                              controllerIdentifier));
+                    handleGroupDoesNotManageByThisController(accountId, groupId);
+                }
+            }
+        }
+        else {
+            LOGGER.error("group locking service failed to get lock for groupId {}, accountId {}.", groupId, accountId);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:844]: print to log dal error and write : staying in the current stat %s

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 15]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+    //    }
+
+    public void syncGroupsOwner() {
+        ILockRepo           lockRepo                    = RepoManager.getInstance().getLockRepo();
+        ApiResponse<String> lockGroupControllerResponse = lockRepo.getLockValueById(groupId, accountId);
+        String              controllerIdentifier        = SpotinstContext.getInstance().getControllerIdentifier();
+
+        if (lockGroupControllerResponse.isRequestSucceed()) {
+            String  lockGroupControllerValue = lockGroupControllerResponse.getValue();
+            boolean isGroupBelongToNone      = lockGroupControllerValue == null;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:825]: isGroupAlreadyBelongToControl = lockGroupControllerValue != null;

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 16]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+            }
+        }
+        else {
+            LOGGER.error("group locking service failed to get lock for groupId {}, accountId {}.", groupId, accountId);
+        }
+    }
+
+    private void handleGroupManagedByNone(String controllerIdentifier) {
+        LOGGER.info(String.format("group %s belong to controller with identifier %s", groupId, controllerIdentifier));
+        GroupStateTracker cloudDetails = SpotinstContext.getInstance().getConnectionStateByGroupId().get(groupId);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:850]: rename to groupAcuiringDetails

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 17]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+                }
+            }
+        }
+        else {
+            LOGGER.error("group locking service failed to get lock for groupId {}, accountId {}.", groupId, accountId);
+        }
+    }
+
+    private void handleGroupManagedByNone(String controllerIdentifier) {
+        LOGGER.info(String.format("group %s belong to controller with identifier %s", groupId, controllerIdentifier));
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:849]: change the log

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 18]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+                    handleGroupDoesNotManageByThisController(accountId, groupId);
+                }
+            }
+        }
+        else {
+            LOGGER.error("group locking service failed to get lock for groupId {}, accountId {}.", groupId, accountId);
+        }
+    }
+
+    private void handleGroupManagedByNone(String controllerIdentifier) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:848]: AcuireLock

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 19]
```diff
@@ -643,6 +728,174 @@ protected Integer getPendingThreshold() {
...
+    //                }
+    //            }
+    //            //there is no controller for the given group in redis, should take ownership
+    //            else {
+    //                handleGroupManagedByThisController(cloud, controllerIdentifier);
+    //            }
+    //        }
+    //    }
+
+    public void syncGroupsOwner() {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:818]: rename to syncGroupOwner

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 11 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:120]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 12 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:820]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 13 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:821]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 14 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:844]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 15 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:825]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 16 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:850]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 17 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:849]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 18 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:848]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 19 above)
Comment (sitay93) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java:818]: done

**Pull request merged**